### PR TITLE
Implement URL Handling from native iOS

### PR DESCRIPTION
**Pull request by MiraiSubject**
This PR exposes the necessary function to handle file opening from external apps. (for example: Safari). 

Am working on an implementation for osu.iOS for handling beatmap imports. 

Provided screenshot is an example of beatmap downloading from safari: 

img_0014
---

Exposes the framework for external file handling. 

Add a sentence or two describing this change in plain english. This will be displayed on the [changelog](https://osu.ppy.sh/home/changelog). A single screenshot or short gif is also welcomed.

### CODE REVIEW

**Comment by peppy** [comment]
Is this even required? can you not just have the code on the osu! side?

**Comment by MiraiSubject** [comment]
Well the framework inherits from UIApplicationDelegate and uses UIKit.
The AppDelegate of the game code inherits from the GameAppDelegate in the framework.

So it made sense to make an abstract method here and implement the actual method in the game. 

See: https://developer.xamarin.com/api/member/MonoTouch.UIKit.UIApplicationDelegate.OpenUrl/ and https://developer.xamarin.com/api/member/MonoTouch.UIKit.UIApplicationDelegate.OpenUrl/p/MonoTouch.UIKit.UIApplication/MonoTouch.Foundation.NSUrl/MonoTouch.Foundation.NSDictionary/

Else you would need to make the AppDelegate of osu!.iOS somehow conform to UIApplicationDelegate but this is not possible in the current structure of the projects.

**Comment by MiraiSubject** [comment]
FileDrop event for iOS now gets fired on the framework's appdelegate. This also exposes the FileDrop method for use in osu.iOS to open files. [ppy/osu#3914 -> AppDelegate.cs](https://github.com/ppy/osu/blob/54c0940447b176bc68c6fcc3f97196d887856d37/osu.iOS/AppDelegate.cs)

File: SampleGame.iOS/AppDelegate.cs [hunk 1]
```diff
@@ -4,12 +4,21 @@
...
 
 namespace SampleGame.iOS
 {
     [Register("AppDelegate")]
     public class AppDelegate : GameAppDelegate
     {
         protected override Game CreateGame() => new SampleGameGame();
+
+
+        override public void fileDrop(object sender, FileDropEventArgs e)
```
Comment (peppy) [inline SampleGame.iOS/AppDelegate.cs]: This isn't how we test. You should made a specific NUnit test if you want to add proper test paths.

File: osu.Framework.iOS/GameAppDelegate.cs [hunk 2]
```diff
@@ -52,5 +55,20 @@ private void aotImageSharp()
...
+            {
+                gameView.OnFileDrop(url.Path);
+            } else if (url.ToString().StartsWith("osu://"))
+            {
+                //TODO: Handle osu URLs for things like direct from the web. 
+            }
+            return true;
+        }
+
+        public abstract void fileDrop(object sender, FileDropEventArgs e);
```
Comment (peppy) [inline osu.Framework.iOS/GameAppDelegate.cs]: public methods are `UpperCamelCase`.

File: osu.Framework.iOS/GameAppDelegate.cs [hunk 3]
```diff
@@ -52,5 +55,20 @@ private void aotImageSharp()
...
             }
         }
+
+
+        public override bool OpenUrl(UIApplication app, NSUrl url, NSDictionary options)
+        {
+            if (url.IsFileUrl) // Built-in check whether Url is actually a file path. 
+            {
+                gameView.OnFileDrop(url.Path);
+            } else if (url.ToString().StartsWith("osu://"))
```
Comment (peppy) [inline osu.Framework.iOS/GameAppDelegate.cs]: this is osu-framework, not osu.

**Comment by peppy** [comment]
Closing for now as I believe you should just start from scratch on this. The history of this issue (and its current state) have many mistakes.

**Pull request closed without merging**